src/Database/LSMTree/Internal/IndexCompactAcc.hs

```diff
@@ -152,6 +152,15 @@ append (AppendMultiPage k n) ica        = appendMulti (k, n) ica
 -- INVARIANTS: see [construction invariants](#construction-invariants).
 appendSingle :: forall s. (SerialisedKey, SerialisedKey) -> IndexCompactAcc s -> ST s (Maybe Chunk)
 appendSingle (minKey, maxKey) ica@IndexCompactAcc{..} = do
+#ifdef NO_IGNORE_ASSERTS
+    lastMinKey <- readSTRef icaLastMinKey
+    let preconditionSorted :: Bool
+        preconditionSorted = minKey <= maxKey && smaybe True (<= minKey) lastMinKey
+    assert preconditionSorted $ pure ()
+    let preconditionPartitioned :: Bool
+        preconditionPartitioned = minRfbits == keyTopBits16 icaRangeFinderPrecision maxKey
+    assert preconditionPartitioned $ pure ()
```

Comment:
I don't think the bindings and type signatures really make it clearer, just feels more noisy. :shrug: 

```suggestion
    assert (minKey <= maxKey && smaybe True (<= minKey) lastMinKey) $ pure ()    -- sorted
    assert (minRfbits == keyTopBits16 icaRangeFinderPrecision maxKey) $ pure ()  -- partitioned
```